### PR TITLE
Worker message plumbing

### PR DESCRIPTION
**Pull request by edman124**
Adding plumbing of autoscaling signals (active_thread_count, maximum_thread_count, active_bundle_count, maximum_bundle_count) to DFE by sending a WorkerMessage with a StreamingScalingReport populated with the mentioned metrics.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by edman124** [comment]
please take an initial look @bvolpato 
I will clean up the logs when finalizing PR

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java [hunk 1]
```diff
@@ -269,4 +277,42 @@ public WorkItemServiceState reportWorkItemStatus(WorkItemStatus workItemStatus)
...
   }
+
+  /** Reports the autoscaling signals to dataflow */
+  @Override
+  public WorkerMessageResponse reportWorkerMessage(StreamingScalingReport report)
+      throws IOException {
+    DateTime endTime = DateTime.now();
+    logger.debug("Reporting WorkMessageResponse");
+    Map<String, String> labels =
+        new HashMap<String, String>(
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:288]: rm new HashMap , that copies entries from the immutable map but then will be mutable.  you can just use the immutable map

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java [hunk 2]
```diff
@@ -269,4 +277,42 @@ public WorkItemServiceState reportWorkItemStatus(WorkItemStatus workItemStatus)
...
+    if (result == null) {
+      logger.warn("Worker Message response is null");
+      throw new IOException("Got null Worker Message response");
+    }
+
+    // Currently no response is expected
+    if (result.getWorkerMessageResponses() == null) {
+      return new WorkerMessageResponse();
+    }
+    WorkerMessageResponse response = result.getWorkerMessageResponses().get(0);
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:314]: verify there is either 1 or 0 responses? this will just ignore more than the first which could be confusing.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 3]
```diff
@@ -1747,6 +1748,14 @@ private void updateThreadMetrics() {
     maxOutstandingBundles.addValue(workUnitExecutor.maximumElementsOutstanding());
   }
 
+  private void sendWorkerMessage() throws IOException {
+    LOG.info("[chengedward] creating StreamingScalingReport");
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1752]: make sure to remove debug lgos

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java [hunk 4]
```diff
@@ -269,4 +277,42 @@ public WorkItemServiceState reportWorkItemStatus(WorkItemStatus workItemStatus)
     logger.debug("ReportWorkItemStatus result: {}", state);
     return state;
   }
+
+  /** Reports the autoscaling signals to dataflow */
+  @Override
+  public WorkerMessageResponse reportWorkerMessage(StreamingScalingReport report)
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:283]: if this stays as is, the method should be renamed.  But I think it would be better if we can combine with other data sent in the same rpc.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 5]
```diff
@@ -1758,6 +1767,13 @@ public void reportPeriodicWorkerUpdates() {
     } catch (Exception e) {
       LOG.error("Unexpected exception while trying to send counter updates", e);
     }
+    try {
+      sendWorkerMessage();
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1771]: do we want to do this for appliance or just streaming engine?

Comment (scwhittle) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1771]: maybe do this from a separate periodic method than the existing method sending the counters?
So that the latency of reporting this message doesn't hold up counters which can affect watermark latency in appliance.

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java (hunk 4 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:283]: I think this is the first WorkerMessage sent in Legacy Worker since its a new path. Is there somewhere else where a similar RPC is sent?

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java (hunk 2 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:314]: Leaving it like this temporarily since this quarter there is work to populate a response and process it on the legacy worker side

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 5 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1771]: Just streaming engine for now. Create a new periodic method to send the WorkerMessage

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java (hunk 4 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:283]: Renamed to reportStreamingMetricsWorkerMessage

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java (hunk 2 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:313]: Removed for now and will add back when there is a response to process

**Comment by edman124** [comment]
Updated PR, should be in a review ready state now

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 6]
```diff
@@ -1747,6 +1757,12 @@ private void updateThreadMetrics() {
     maxOutstandingBundles.addValue(workUnitExecutor.maximumElementsOutstanding());
   }
 
+  private void sendWorkerMessage() throws IOException {
```
Comment (zuxie) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1760]: Can you populate other processed signals here as well? Namely maximum_thread_count, active_bundle_count, maximum_bundle_count.

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java [hunk 7]
```diff
@@ -280,13 +279,12 @@ public WorkItemServiceState reportWorkItemStatus(WorkItemStatus workItemStatus)
 
   /** Reports the autoscaling signals to dataflow */
   @Override
-  public WorkerMessageResponse reportWorkerMessage(StreamingScalingReport report)
+  public WorkerMessageResponse reportStreamingMetricsWorkerMessage(StreamingScalingReport report)
```
[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:282]: Done. Went with the second method

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java [hunk 8]
```diff
@@ -311,8 +309,6 @@ public WorkerMessageResponse reportWorkerMessage(StreamingScalingReport report)
     if (result.getWorkerMessageResponses() == null) {
       return new WorkerMessageResponse();
     }
-    WorkerMessageResponse response = result.getWorkerMessageResponses().get(0);
-    logger.debug("Worker Message Response result: {}", response);
-    return response;
+    return new WorkerMessageResponse();
```
[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClient.java:312]: Done. Made the method void for now.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 9]
```diff
@@ -587,6 +588,13 @@ public void start() {
         options.getWindmillHarnessUpdateReportingPeriod().getMillis(),
         TimeUnit.MILLISECONDS);
 
+    workerMessageReportTimer = executorSupplier.apply("WorkerMessageTimer");
+    workerMessageReportTimer.scheduleWithFixedDelay(
+        this::reportPeriodicWorkerMessage,
+        0,
+        options.getWindmillHarnessUpdateReportingPeriod().getMillis(),
```
[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:595]: What option would I check against? This also seems more similar to the periodicWorkerUpdates which has no check

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClientTest.java [hunk 10]
```diff
@@ -227,10 +230,22 @@ public void testCloudServiceCallMultipleWorkItems() throws Exception {
...
+    SendWorkerMessagesResponse workerMessage = new SendWorkerMessagesResponse();
+    workerMessage.setFactory(Transport.getJsonFactory());
+    response.setContent(workerMessage.toPrettyString());
+    when(request.execute()).thenReturn(response);
+    StreamingScalingReport activeThreadsReport =
+        new StreamingScalingReport().setActiveThreadCount(1);
+    WorkUnitClient client = new DataflowWorkUnitClient(pipelineOptions, LOG);
+    WorkerMessageResponse expected = new WorkerMessageResponse();
+    WorkerMessageResponse clientResponse =
+        client.reportStreamingMetricsWorkerMessage(activeThreadsReport);
```
[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClientTest.java:246]: Done.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 6 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:1760]: Done.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (scwhittle) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:595]: only schedule here if options.getWindmillHarnessUpdateReportingPeriod() > 0

don't schedule with fixed delay.  Not sure if it would throw exception or just run in tight loop but neither seems particularly desirable and having a way to disable if desired for some reason could be nice.

[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:595]: A little confused on the part about not scheduling with fixed delay, the other scheduling option is with fixed period since we need the metrics to be sent every so often like the counter metrics being sent.

[the remark this answers is not in the document]
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: updated to add check before scheduling, but still using schedule with fixed delay, which seems to be working without any problems

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkUnitClient.java [hunk 11]
```diff
@@ -49,4 +51,12 @@ interface WorkUnitClient {
    * @return a {@link WorkItemServiceState} (e.g. a new stop position)
    */
   WorkItemServiceState reportWorkItemStatus(WorkItemStatus workItemStatus) throws IOException;
+
+  WorkerMessage createWorkerMessageFromStreamingScalingReport(StreamingScalingReport report);
+  /**
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkUnitClient.java:56]: blank line above

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 12]
```diff
@@ -252,8 +253,8 @@ public class StreamingDataflowWorker {
   private final Counter<Integer, Integer> totalAllocatedThreads;
   private final Counter<Long, Long> outstandingBytes;
   private final Counter<Long, Long> maxOutstandingBytes;
-  private final Counter<Long, Long> outstandingBundles;
-  private final Counter<Long, Long> maxOutstandingBundles;
+  private final Counter<Integer, Integer> outstandingBundles;
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:256]: I'm not sure if dataflow supports changing counter type in this way.  Additionally the max may be set higher (effectively unbounded) to some value over Integer.MAX currently.  It seems safer for that reason to revert the change from Long->Integer and instead just bound to int range when populating the scaling proto (presuming that is what motivated this change).

File: runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClientTest.java [hunk 13]
```diff
@@ -227,6 +231,26 @@ public void testCloudServiceCallMultipleWorkItems() throws Exception {
     client.getWorkItem();
   }
 
+  @Test
+  public void testReportWorkerMessageEmptyResponse() throws Exception {
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/DataflowWorkUnitClientTest.java:235]: nit: remove empty response from test name

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 12 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:256]: I reverted the Long->Integer change and just cast the value from the BoundedQueueExecutor as a Long since it is stored as an int under the hood.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 14]
```diff
@@ -586,6 +588,15 @@ public void start() {
         options.getWindmillHarnessUpdateReportingPeriod().getMillis(),
         TimeUnit.MILLISECONDS);
 
+    workerMessageReportTimer = executorSupplier.apply("WorkerMessageTimer");
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:591]: this needs to be shutdown and awaited in stop() below like the other timers.

You could consider refactoring so that we have a collection of these timers that we add to after we start them, since the shutdown/await code is repetitive and I don't think we otherwise refer to them.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkUnitClient.java [hunk 15]
```diff
@@ -49,4 +51,17 @@ interface WorkUnitClient {
    * @return a {@link WorkItemServiceState} (e.g. a new stop position)
    */
   WorkItemServiceState reportWorkItemStatus(WorkItemStatus workItemStatus) throws IOException;
+  /**
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkUnitClient.java:54]: blank line above to separate from previous method (I'm suprised spotlessApply doesn't care but might as well match the rest of the file).

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 14 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Refactored and fixed.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 16]
```diff
@@ -579,72 +578,91 @@ public void start() {
...
+    timersMap.put("GlobalWorkerUpdatesTimer", executorSupplier.apply("GlobalWorkerUpdates"));
+    timersMap
+        .get("GlobalWorkerUpdatesTimer")
+        .scheduleWithFixedDelay(
+            this::reportPeriodicWorkerUpdates,
+            0,
+            options.getWindmillHarnessUpdateReportingPeriod().getMillis(),
+            TimeUnit.MILLISECONDS);
+
+    timersMap.put("WorkerMessageTimer", executorSupplier.apply("ReportWorkerMessage"));
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:590]: no need to create this executor if not actually scheduling anything on it, see above.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java [hunk 17]
```diff
@@ -276,13 +277,11 @@ public class StreamingDataflowWorker {
   private final HotKeyLogger hotKeyLogger;
   // Periodic sender of debug information to the debug capture service.
   private final DebugCapture.@Nullable Manager debugCaptureManager;
-  private ScheduledExecutorService refreshWorkTimer;
-  private ScheduledExecutorService statusPageTimer;
-  private ScheduledExecutorService globalWorkerUpdatesTimer;
+  // Map from timer name to ScheduledExecutorService
+  private ConcurrentMap<String, ScheduledExecutorService> timersMap = new ConcurrentHashMap();
```
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java:281]: I think that timersMap is a little confusing given the windmill Timer terminology.
How about just scheduledExecutors? or periodicFunctionExecutors?

Also as a nit, we're not using the key or using this concurrently and it makes things a little more complicated. I think you could change this to an ArrayList and avoid repeating the key. You can also only create the executor if needed.

Example:
```
if (...) {
  ScheduledExecutorService workUpdatesService = 
     executorSupplier.apply("GlobalWorkerUpdates");
  workUpdatesService.scheduleWithFixedDelay();
  scheduledExecutors.add(workUpdateService);
}
```

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 17 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Done.

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java (hunk 16 above)
Comment (edman124) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java]: Done.

**Review by scwhittle — APPROVED** [review]
